### DISCUSSION TITLE
Why are Prisma generated types in the `select` object of a `create` disregarded?

### DISCUSSION THREAD

**Discussion by mannalext** [prisma/prisma#18769] [Q&A] [answered]
I'm in the process of implementing Prisma across my project. I came across an issue when trying to use a generated type. I'm trying to do a simple create and dictate that only the `id` of the resulting object be returned. The following is a small example of what I mean:

```
    const leagueId: Prisma.leaguesSelect = {
      id: true,
    }

    const result = await prisma.leagues.create({
      data: {
        name: leagueName,
      },
      select: leagueId,
    });

    return result.id;
```

However, TypeScript complains about the return statement, implying that `result.id` could be undefined (the function signature returns a number). At first I thought this was because Prisma wasn't guaranteeing to get anything back from the database and wouldn't throw an error if that was the case, but when I try the following:

```
    const result = await prisma.leagues.create({
      data: {
        name: leagueName,
      },
      select: { id: true },
    });

    return result.id;
```

The compilation error disappears. It's almost as if the `select` object is disregarding my instantiated `Prisma.leaguesSelect`. I have to be explicit to get the behavior I want. 

I can't imagine that this is a bug with Prisma because what I'm trying to accomplish is pretty mundane. Can someone give some insight about where I might be going wrong?

**Comment by nurul3101** [comment, marked as answer]
Hello @mannalext 👋 

The issue you are encountering seems to be due to a slight misunderstanding in how the `Prisma.leaguesSelect` type should be used. The `Prisma.leaguesSelect` type is a generated type by Prisma based on your schema, and it represents an object with possible selections for your leagues model. The keys of the object are the fields in the model, and the values are booleans indicating whether the field should be selected or not.

However, when you create a `leagueId` object with the type `Prisma.leaguesSelect`, you are not enforcing that the `id` property is always true. TypeScript sees that you have specified leagueId as an instance of `Prisma.leaguesSelect`, so it understands that id can be either true or false. That's why you get the error complaining about the possibility of result.id being undefined.

You could create a type that enforces the presence of the `id` field, one solution could be
```ts
type LeagueIdSelect = {
  id: true;
} & Omit<Prisma.LeaguesSelect, "id">;

const leagueId: LeagueIdSelect = {
  id: true,
};

const result = await prisma.leagues.create({
  data: {
    name: leagueName,
  },
  select: leagueId,
});

return result.id;
```

This new `LeagueIdSelect` type is defined as an object with an enforced `id: true` property and all the other properties from `Prisma.LeaguesSelect`, but without the id key.

This way, TypeScript will understand that the id property will always be present in the result object, and you won't get the error.

Please let me know in case you face any issues with this approach.

**Reply by mannalext** [reply]
This makes sense and helps quite a bit, thank you. I took another look at the documentation after reading your response and realized I was misinterpreting how to best use the `select` object. Thank you very much for taking the time to reply!
